File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.py
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        ch = dict()
        l =0
        longest = 0
        for r in range(len(s)):
            if s[r] in ch:
                l = max(l,ch[s[r]]+1)
            ch[s[r]] = r
            longest = max(longest,(r-l+1))
        return longest  
                
                

File: 3-longest-substring-without-repeating-characters/longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int maxLen=0,start=0;
        HashMap<Character,Integer> mp = new HashMap<>();
        for(int i=0;i<s.length();i++){
            if(mp.containsKey(s.charAt(i)) && mp.get(s.charAt(i))>=start){
                start = mp.get(s.charAt(i))+1;
            }
            mp.put(s.charAt(i),i);
            maxLen = Math.max(maxLen,i-start+1);
        }
        return maxLen;
    }
}